src/main/java/org/apache/commons/lang3/ThreadUtils.java

```diff
@@ -0,0 +1,341 @@
+            throw new IllegalArgumentException("The threadGroupName must not be null");
+        }
+
+        final Thread thread = findThreadById(threadId);
+
+        if(thread != null && thread.getThreadGroup() != null && thread.getThreadGroup().getName().equals(threadGroupName)) {
+            return thread;
+        }
+        return null;
+    }
+
+    /**
+     * Return active threads with the specified name if they belong to a specified thread group
+     *
+     * @param threadName The thread name
+     * @param threadGroupName The thread group
+     * @return The threads which belongs to a thread group and the thread's name match the specified name,
+     * An empty array is returned if no such thread exists
+     * @throws IllegalArgumentException if the specified thread name or group is null
+     * @throws  SecurityException
+     *          if the current thread cannot access the system thread group
+     *
+     * @throws  SecurityException  if the current thread cannot modify
+     *          thread groups from this thread's thread group up to the system thread group
+     */
+    public static Thread[] findThreadsByName(final String threadName, final ThreadGroup threadGroup) {
+        if (threadName == null) {
+            throw new IllegalArgumentException("The threadName must not be null");
+        }
+        if (threadGroup == null) {
+            throw new IllegalArgumentException("The threadGroupName must not be null");
+        }
+
+        final Thread[] threads = findThreadsByName(threadName);
+
+        if(threads.length == 0) {
+            return EMPTY_THREAD_ARRAY;
+        }
+
+        final List<Thread> matchedThreads = new ArrayList<Thread>();
+
+        for (int i = 0; i < threads.length; i++) {
+            final Thread thread = threads[i];
+
+            if(thread != null && threadGroup.equals(thread.getThreadGroup())) {
+                matchedThreads.add(thread);
+            }
+
+        }
+        return matchedThreads.toArray(new Thread[matchedThreads.size()]);
+    }
+
+    /**
+     * Return active threads with the specified name if they belong to a thread group with the specified group name
+     *
+     * @param threadName The thread name
+     * @param threadGroupName The thread group name
+     * @return The threads which belongs to a thread group with the specified group name and the thread's name match the specified name,
+     * An empty array is returned if no such thread exists
+     * @throws IllegalArgumentException if the specified thread name or group name is null
+     * @throws  SecurityException
+     *          if the current thread cannot access the system thread group
+     *
+     * @throws  SecurityException  if the current thread cannot modify
+     *          thread groups from this thread's thread group up to the system thread group
+     */
+    public static Thread[] findThreadsByName(final String threadName, final String threadGroupName) {
```

Comment:
i think looping over thread groups and then over threads is overall more efficient, because it removes more candidates earlier. but I would write a JMH bench / check the iteration code in enumerate(T/TG) to back this assumption